Add menu plugin docs (#81)
grammyjs/website: site/docs/plugins/menu.md

</CodeGroupItem>
</CodeGroup>

Naturally, if you are using a custom context type, you can pass it to `Menu`, too.

```ts
const menu = new Menu<MyContext>("id");
```

## Adding Buttons

> The menu plugin layouts your keyboards exactly like [the plugin for inline keyboards](./keyboard.md#building-an-inline-keyboard) does.

Author: Oops

It will not work in other middleware elsewhere on your bot because it is not clear which menu should update.

```ts
const menu = new Menu("time")
  .text(
    () => new Date().toString(), // button label is current time
    (ctx) => ctx.menu.update(), // update time on button click
  );
```

You can also update the menu implicitly by editing the corresponding message.

```ts
const menu = new Menu("time")
  .text(
    () => new Date().toString(),
    (ctx) => ctx.editMessageText("Last updated: " + new Date.toString()),
  );
```

The menu will detect that you edit the text of the message, and use the opportunity to update the buttons uderneath, too.
As a result, you can often skip the `ctx.menu.update()` calls.

Calling `ctx.menu.update()` does not update the menu immediately.
Instead, it only sets a flag and remembers to update it at some point during your middleware execution.
This is called _lazy updating_.
If you edit the message itself later on, the plugin can simply use the same API call to also update the buttons.
This is very efficient, and it makes sure that both the message and the keyboard are updated at the same time.

> Naturally, if you called `ctx.menu.update()` but you never perform any edits to the message, the menu plugin makes sure to update your buttons before you middleware completes.

Reviewer: It refers to "the buttons" in this case, so "by themselves" is correct.

Maybe this is better though, and lets us use itself:
```suggestion
Naturally, if you call `ctx.menu.update()` but you never request any edits to the message, the menu plugin will update the keyboard by itself, before your middleware completes.
```

You can set a different parent, too.

```ts
// Register settings menu at main menu
main.register(settings);
// Optionally, set a different parent
main.register(settings, "back-from-settings-menu");
```

You can register as many menus as you like, and nest them as deeply as you like.
The menu identifiers let you jump easily to any page.

**Note that you only have to make a single menu of your structure interactive.**
For example, only pass the root menu to `bot.use`.

```ts
// Do this:
bot.use(main);

// Don't do this:
bot.use(main);
bot.use(settings);
```

## Payloads

You can store short texts along with all text and navigation buttons.
When the respective handlers are invoked, the payload will be available under `ctx.match`.
This is useful because it lets you store a little bit of data in a menu.

Here is an example menu that remembers its creator in the payload.
Other use cases could for example be to store the index in a pagination.

```ts
const menu = new Menu("pun-intended")
  .text(
    { text: "I know my creator", payload: (ctx) => ctx.from.first_name },
    (ctx) => ctx.reply(`I was once created by ${ctx.match}!`),
  );

bot.use(menu);
bot.command("menu", async (ctx) => {

>   await ctx.reply("You summoned a demenu!", { reply_markup: menu });

Reviewer: Nope, I can't. 😔

>   await ctx.reply("You summoned a demenu!", { reply_markup: menu });

Author: That's fine, I'll just remove it.

});
```

Payloads also work well together with dynamic ranges.

## Dynamic Ranges

So far, we've only seen how to change the text on a button dynamically.
You can also dynamically adjust the structure of a menu in order to add and remove buttons on the fly.

::: danger Never Change a Menu During Message Handling
You cannot create or change menus your during message handling.

> All menus must be fully created and registered before your bot starts.

Reviewer: In that case, let's avoid the "should" and treat it as if it is not possible at all.

So, the first block of suggested changes.

Adding new menus while your bot is running would cause a memory leak.
Your bot would slow down more and more, and eventually crash.
:::

You can let a part of a menu's buttons be generated on the fly (or all of them if you want).
We call this part of the menu a _dynamic range_.
The easiest way to create a dynamic range is by using the `MenuRange` class that this plugin provides.
A `MenuRange` provides you with exactly the same functions as a menu, but it does not have an identifier, and it cannot be registered.

```ts
function getRandomInt(min: number, max: number) {

>   return min + Math.floor(Math.random() * (max - min));

Reviewer: Oh, okay... I guess this is considered as "normal" so will be assumed to work that way?

To me it seems like really unexpected behavior. Maybe we can add a comment to avoid this issue:
```ts
// Returns a random number between and min [inclusive] and max [exclusive] 
```

>   return min + Math.floor(Math.random() * (max - min));

Author: Yes, that's default behaviour. Check python's `range`, and `String.prototype.substring` in JS, and all other things in all languages.
Note also that `Math.random()` itself returns values from `[0, 1)`. We just reflect that. In other words, if the 1 would be inclusive, `max` would automatically be inclusive, too.

I'm okay with adding a comment if that's helpful. I'm only hesitant because random numbers are not the point of the example.

>   return min + Math.floor(Math.random() * (max - min));

Author: How about calling the arguments `minInclusive` and `maxExclusive`?

>   return min + Math.floor(Math.random() * (max - min));

Author: Please have a look at 40a13a6

>   return min + Math.floor(Math.random() * (max - min));

Reviewer: That's perfect and much better than a comment.

I was originally going to suggest a change like this, but thought it would be less likely to be accepted than just adding a comment.

}

// Create a menu with a random number of buttons
const menu = new Menu("random");

menu.dynamic((_ctx) => {
  const range = new MenuRange();
  const buttonCount = getRandomInt(2, 10);
  for (let i = 0; i < buttonCount; i++) {
    range
      .text(i.toString(), (ctx) => ctx.reply(`${i} selected`))
      .row();
  }
  return range;
});

menu.text("Generate new", (ctx) => ctx.menu.update());
```

> The range builder function that you pass to `dynamic` may be `async`, so you can even perform API calls or do database communication.

Author: This is not a menu! It's a menu range. I'm trying to make it very clear what the difference is. We could say “before returning the generated menu range”?

> The range builder function that you pass to `dynamic` may be `async`, so you can even perform API calls or do database communication.

Reviewer: Yes, "menu range" works just as well here.

Moreover, the range builder function takes a context object as a first argument.
(This is unused in the example above).

Optionally, as a second argument after `ctx`, you can receive a fresh instance of `MenuRange`.
You can modify it instead of returning your own instance if that's what you prefer.

```ts
menu.dynamic((ctx, range) => {
  for (const text of ["foo", "bar", "baz"]) {
    range // no need for `new MenuRange()` or a `return`
      .text(text, (ctx) => ctx.reply(text))
      .row();
  }
});
```

## Answering Callback Queries Manually

The menu plugin will call `answerCallbackQuery` automatically for its own buttons.
You can set `autoAnswer: false` if you want to disable this.

```ts
const menu = new Menu("id", { autoAnswer: false });
```

You will now have to call `answerCallbackQuery` yourself.
This allows you to pass custom messages that are displayed to the user.

## Outdated Menus and Fingerprints

Let's say you have a menu where a user can toggle notifications on and off, such as in the example [up here](#dynamic-labels).

> Now, if a user sends e.g. `/settings` twice, they will get the same menu twice.

Reviewer: ```suggestion
Now, if a user sends `/settings` twice, they will get the same menu twice.
```

> But changing the notification setting on one of the two messages will not update the other!

Reviewer: ```suggestion
But, changing the notification setting on one of the two messages will not update the other!
```

> It is clear that we cannot keep track of all settings messages in a chat, and update all old menus in the entire chat history.

Reviewer: ```suggestion
It is clear that we cannot keep track of all settings messages in a chat, and update all old menus across the entire chat history.
```

> You would have to use so many API calls for this that Telegram will restrict your bot.

Reviewer: ```suggestion
You would have to use so many API calls for this that Telegram would rate-limit your bot.
```

> You also need to buy a lot of storage to remember all the message identifiers of all menus in all chats.

Reviewer: ```suggestion
You would also require a lot of storage to remember all of the message identifiers of every menu, across all chats.
```

> That is not practical.

Reviewer: ```suggestion
This is not practical.
```

> The solution to this is to check if a menu is outdated before performing any actions.

Reviewer: ```suggestion
The solution, is to check if a menu is outdated _before_ performing any actions.
```

Author: Are you sure about the comma? Otherwise I agree

Reviewer: The text is better with it, but will not be technically incorrect without it.

> That means that we will only update old menus if the user actually starts clicking any buttons on them.

Reviewer: ```suggestion
This way, we will only update old menus if a user actually starts clicking the buttons on them.
```

> The menu plugin does this automatically for you, so you don't need to worry about it.

Reviewer: ```suggestion
The menu plugin handles this automatically for you, so you don't need to worry about it.
```

> You can configure what happens when an outdated menu is detected.

Reviewer: Since there is full control, and not a choice between presets.
```suggestion
You can configure exactly what happens when an outdated menu is detected.
```

By default, the message “Menu was outdated, try again!” will be displayed to the user, and the menu will be updated.
You can define custom behavior in the config under `onMenuOutdated`.

```ts
// Custom message to be displayed
const menu0 = new Menu("id", { onMenuOutdated: "Updated, try now." });
// Custom handler function
const menu1 = new Menu("id", {
  onMenuOutdated: async (ctx) => {
    await ctx.answerCallbackQuery();
    await ctx.reply("Here is a fresh menu", { reply_markup: menu1 });
  },
});
// Completely disable outdated check (may run wrong button handlers)
const menu2 = new Menu("id", { onMenuOutdated: false });
```

We have a heuristic to check if the menu is outdated.

> It looks at

Reviewer: ```suggestion
It looks at:
```

- the identifier of the menu,
- the shape of the menu,
- the position of the pressed button,
- the payload, if specified, and
- the text of the pressed button.

This data is effeciently compressed into a 4-byte hash that is stored in every button.
It will then be compared to see if the menu is outdated before any handler is run.

> It is thinkable that your menu changes, but all of the above things are staying the same.

Reviewer: ```suggestion
It is possible that your menu changes, while all of the above things stay the same.
```

> This is usually not the case, but if you are creating a menu that has this property, you should use a fingerprint function.

Reviewer: ```suggestion
This is usually not the case, but if you are creating a menu where this is likely, you should use a fingerprint function.
```

```ts
function ident(ctx: Context): string {
  // return string that would change when your menu changes
}
const menu = new Menu("id", { fingerprint: (ctx) => ident(ctx) });
```

The fingerprint string will be added to the above list of things that impact the hash generation.

> That way, you can make sure that outdated menus are always detected.

Reviewer: ```suggestion
This way, you can be sure that outdated menus are always detected.
```

## How Does It Work

The menu plugin works completely without storing any data.
This is important for large bots with millions of users.

> Saving the states of all menus would consume too much memory.

Reviewer: ```suggestion
Saving the state of all menus would consume too much memory.
```

When you create your menu objects and link them together via `register` calls, no menus are actually built.
Instead, the menu plugin will remember how to assemble new menus based on your operations.

> Whenever a menus is sent, it will replay these operation to render your menu.

Reviewer: ```suggestion
Whenever a menu is sent, it will replay these operations to render your menu.
```

> This includes layouting all dynamic ranges and generating all dynamic labels.

Reviewer: ```suggestion
This includes laying out all dynamic ranges and generating all dynamic labels.
```

Once the menu is sent, the rendered button array will be forgotten again.

> When a user presses a button of the menu, we need to find the handler that was added to that button at the time the menu was rendered.

Reviewer: ```suggestion
When a user presses a menu's button, we need to find the handler that was added to that button at the time the menu was rendered.
```

Hence, we simply render the old menu again.
However, this time, we don't actually need the full layout—all we need is the overall structure, and that one specific button.
Consequently, the menu plugin will perform a shallow rendering in order to be more efficient.
In other words, the menu will only be rendered partially.

Once the pressed button is known again (and we have checked that the menu is not [outdated](#outdated-menus-and-fingerprints)), we invoke the handler.

> Internally, the menu plugin makes heavy use of [API Transformer Functions](/advanced/transformers.md), e.g. to quickly render outgoing menus on the fly.

Reviewer: ```suggestion
Internally, the menu plugin makes heavy use of [API Transformer Functions](/advanced/transformers.md), for example, to quickly render outgoing menus on the fly.
```

> When you register the menus in a large hierachy for navigating, they will in fact not store these references explicitly.

Reviewer: ```suggestion
When you register the menus in a large navigation hierarchy, they will in fact not store these references explicitly.
```

Under the hood, all menus of that one structure are added to the same large pool, and that pool is shared across all contained instances.
Every menu is responsible for every other one in the index, and they can handle and render each other.

> (Most often, it is only the root menu that is actually passed to `bot.use` and receiving any updates.

Reviewer: ```suggestion
(Most often, it is only the root menu that is actually passed to `bot.use` and that receives any updates.
```

> In that case, this one instance will handle the complete pool.)

Reviewer: ```suggestion
In such cases, this one instance will handle the complete pool.)
```

> As a result, you are able to navigate between arbitrary menus without limits, and still the update handling can happen in `O(1)` because there is no need to search through hierarchies after the right menu to handle a button click.

Reviewer: Not sure how ` are handled in hyperlinks, will this work?
```suggestion
As a result, you are able to navigate between arbitrary menus without limit, all while the update handling can happen in [`O(1)` time complexity](https://en.wikipedia.org/wiki/Time_complexity#Constant_time) because there is no need to search through entire hierarchies to find the right menu to handle any given button click.
```

Author: Yes.
